### PR TITLE
feat: remove run context

### PR DESCRIPTION
**Pull request by Eikix**
- remove run context, merge it into virtual machine
- add comments on instructions
- replaced enums by strongly typed string literals

### CODE REVIEW

File: src/runners/cairoRunner.ts [hunk 1]
```diff
@@ -43,23 +43,21 @@ export class CairoRunner {
 
   // Run until the given PC is reached.
   runUntilPc(finalPc: Relocatable, verbose: boolean = false): void {
-    while (this.vm.runContext.pc.getOffset() < finalPc.getOffset()) {
+    while (this.vm.pc.getOffset() < finalPc.getOffset()) {
       this.vm.step();
       if (verbose) {
         console.log(
-          `AP: ${this.vm.runContext.ap.getSegmentIndex()}:${this.vm.runContext.ap.getOffset()}`
+          `AP: ${this.vm.ap.getSegmentIndex()}:${this.vm.ap.getOffset()}`
```
Comment (ClementWalter) [inline src/runners/cairoRunner.ts:50]: we could have a `toString()` for relocatable

File: src/runners/cairoRunner.ts [hunk 2]
```diff
@@ -43,23 +43,21 @@ export class CairoRunner {
...
         console.log(
-          `PC: ${this.vm.runContext.pc.getSegmentIndex()}:${this.vm.runContext.pc.getOffset()}`
+          `PC: ${this.vm.pc.getSegmentIndex()}:${this.vm.pc.getOffset()}`
         );
 
-        console.log(
-          `[ap - 1]: ${this.vm.memory.get(this.vm.runContext.ap.sub(1))}`
-        );
-        console.log(`[fp]: ${this.vm.memory.get(this.vm.runContext.fp)}`);
+        console.log(`[ap - 1]: ${this.vm.memory.get(this.vm.ap.sub(1))}`);
```
Comment (ClementWalter) [inline src/runners/cairoRunner.ts:59]: is there a better logger for ts?

File: src/vm/instruction.ts [hunk 3]
```diff
@@ -31,74 +31,39 @@ import { UnsignedInteger } from 'primitives/uint';
...
 
-// Pc Update
-export enum PcUpdate {
-  Regular = 0,
-  Jump = 1,
-  JumpRel = 2,
-  Jnz = 4,
-}
+export type PcUpdate =
+  | 'no-op'
```
Comment (ClementWalter) [inline src/vm/instruction.ts:41]: pc = pc

File: src/vm/instruction.ts [hunk 4]
```diff
@@ -31,74 +31,39 @@ import { UnsignedInteger } from 'primitives/uint';
...
+export type ApUpdate = 'no-op' | 'ap = ap + res' | 'ap++' | 'ap += 2';
 
-// Fp Update
-export enum FpUpdate {
-  Regular = 0,
-  ApPlus2 = 1,
-  Dst = 2,
-}
+export type FpUpdate =
+  | 'no-op'
```
Comment (ClementWalter) [inline src/vm/instruction.ts:49]: fp = fp

File: src/vm/instruction.ts [hunk 5]
```diff
@@ -31,74 +31,39 @@ import { UnsignedInteger } from 'primitives/uint';
...
-}
+export type Opcode = 'no-op' | 'call' | 'return' | 'assert_eq';
 
 export class Instruction {
-  // The offset to add or sub to ap/fp to obtain the Destination Operand
-  public offDst: number;
-  public offOp0: number;
-  public offOp1: number;
+  public dstOffset: number;
+  public operandZeroOffset: number;
```
Comment (ClementWalter) [inline src/vm/instruction.ts:57]: op0Offset

File: src/vm/instruction.ts [hunk 6]
```diff
@@ -31,74 +31,39 @@ import { UnsignedInteger } from 'primitives/uint';
...
+  public operandOneOffset: number;
```
Comment (ClementWalter) [inline src/vm/instruction.ts:58]: op1Offset

File: src/vm/instruction.ts [hunk 7]
```diff
@@ -31,74 +31,39 @@ import { UnsignedInteger } from 'primitives/uint';
...
   // The register to use as the Destination Operand
-  public dstReg: DstRegister;
+  public dstRegister: RegisterFlag;
   // The register to use as the Operand 0
-  public op0Reg: Op0Register;
+  public operandZeroRegister: RegisterFlag;
```
Comment (ClementWalter) [inline src/vm/instruction.ts:62]: op0Register

File: src/vm/instruction.ts [hunk 8]
```diff
@@ -31,74 +31,39 @@ import { UnsignedInteger } from 'primitives/uint';
...
   // The source of the Operand 1
-  public op1Src: Op1Src;
+  public operandOneSource: OperandOneSource;
```
Comment (ClementWalter) [inline src/vm/instruction.ts:64]: op1Register

File: src/vm/instruction.ts [hunk 9]
```diff
@@ -31,74 +31,39 @@ import { UnsignedInteger } from 'primitives/uint';
...
   // The result logic
-  public resLogic: ResLogic;
+  public resultLogic: ResultLogic;
```
Comment (ClementWalter) [inline src/vm/instruction.ts:66]: opLogic

File: src/vm/instruction.ts [hunk 10]
```diff
@@ -113,84 +78,66 @@ export class Instruction {
...
-    resLogic: ResLogic,
+    desReg: RegisterFlag,
+    operandZeroRegister: RegisterFlag,
+    operandOneSource: OperandOneSource,
+    resultLogic: ResultLogic,
     pcUpdate: PcUpdate,
     apUpdate: ApUpdate,
     fpUpdate: FpUpdate,
     opcode: Opcode
   ) {
```
Comment (ClementWalter) [inline src/vm/instruction.ts:104]: dstOffset
op0Offset
op1Offset
dstRegister
op0Register
op1Source // or op1Register

File: src/vm/instruction.ts
Comment (ClementWalter) [inline src/vm/instruction.ts:1]: we need to stick to one convention regarding prefix or suffix, and it seems like we go for dstOffset, op0Offset, etc.

File: src/vm/instruction.ts [hunk 11]
```diff
@@ -113,84 +78,66 @@ export class Instruction {
...
+    // See Cairo whitepaper, page 32 - https://eprint.iacr.org/2021/1063.pdf.
+    // ┌─────────────────────────────────────────────────────────────────────────┐
+    // │                     off_dst (biased representation)                     │
+    // ├─────────────────────────────────────────────────────────────────────────┤
+    // │                     off_op0 (biased representation)                     │
+    // ├─────────────────────────────────────────────────────────────────────────┤
+    // │                     off_op1 (biased representation)                     │
+    // ├─────┼─────┼───┬───┬───┼───┬───┼───┬───┬───┼────┬────┼────┬────┬────┼────┤
+    // │  0  │  1  │ 2 │ 3 │ 4 │ 5 │ 6 │ 7 │ 8 │ 9 │ 10 │ 11 │ 12 │ 13 │ 14 │ 15 │
+    // └─────┴─────┴───┴───┴───┴───┴───┴───┴───┴───┴────┴────┴────┴────┴────┴────┘
```
Comment (ClementWalter) [inline src/vm/instruction.ts:138]: one time this comment in the file is enough

File: src/runners/cairoRunner.ts (hunk 1 above)
Comment (Eikix) [reply src/runners/cairoRunner.ts:50]: true

Comment (Eikix) [reply src/runners/cairoRunner.ts:50]: adding to the new issue

File: src/runners/cairoRunner.ts (hunk 2 above)
Comment (Eikix) [reply src/runners/cairoRunner.ts:59]: will investigate, and create a tracer epic

**Pull request merged**